scripts/NuGetUtilities.psm1

```diff
@@ -0,0 +1,155 @@
+#>
+
+Set-StrictMode -Version Latest
+$ErrorActionPreference = "Stop"
+$InformationPreference = "Continue"
+
+Import-Module $PSScriptRoot\ScriptUtilities.psm1
+Import-Module $PSScriptRoot\Projects.psm1
+
+$NugetExePath = "$RepoRoot\.nuget\NuGet.exe"
+$NuGetPackageRoot = "$SourceRoot\packages"
+$PackageSource = "https://nuget.org"
+$PackageOutputDirectoryRoot = Join-Path $BinRoot NuGet
+
+function Get-PackageVersion([switch]$previous) {
+    $versionPrefix, $versionSuffix = & $PSScriptRoot\Get-VersionConstants.ps1 -Previous:$previous
+    $version = $versionPrefix
+    if ($versionSuffix)
+    {
+        $version += "-$versionSuffix"
+    }
+
+    $version
+}
+
+function Get-PackageDirectoryName($configuration) {
+    Join-Path $PackageOutputDirectoryRoot $configuration
+}
+
+function New-NuGetPackageFromProjectFile($configuration, $project, $version) {
+    $projectFile = "$SourceRoot\$project\$project.csproj"
+
+    $arguments =
+        "pack", $projectFile,
+        "--configuration", $configuration,
+        "--no-build", "--no-restore",
+        "--include-source", "--include-symbols",
+        "-p:Platform=$Platform",
+        "--output", (Get-PackageDirectoryName $configuration)
+
+    Write-CommandLine dotnet $arguments
+
+    dotnet $arguments
+    if ($LASTEXITCODE -ne 0) {
+        Exit-WithFailureMessage $ScriptName "$project NuGet package creation failed."
+    }
+}
+
+function New-NuGetPackageFromNuSpecFile($configuration, $project, $version, $suffix = "") {
+    $nuspecFile = "$SourceRoot\NuGet\$project.nuspec"
+
+    $arguments=
+        "pack", $nuspecFile,
+        "-Symbols",
+        "-Properties", "platform=$Platform;configuration=$configuration;version=$version",
+        "-Verbosity", "Quiet",
+        "-BasePath", ".\",
+        "-OutputDirectory", (Get-PackageDirectoryName $configuration)
+
+    if ($suffix -ne "") {
+        $arguments += "-Suffix", $Suffix
+    }
+
+    Write-CommandLine $NuGetExePath $arguments
+
+    & $NuGetExePath $arguments
+    if ($LASTEXITCODE -ne 0) {
+        Exit-WithFailureMessage $ScriptName "$project NuGet package creation failed."
+    }
+
+    Write-Information "  Successfully created package '$BinRoot\NuGet\$Configuration\$Project.$version.nupkg'."
+}
+
+function New-NuGetPackages($configuration, $projects) {
+    $version = Get-PackageVersion
+
+    # We can build the NuGet packages for library projects directly from their
+    # project file.
+    foreach ($project in $projects.NewLibrary) {
+        New-NuGetPackageFromProjectFile $configuration $project $version
+    }
+
+    # Unfortunately, application projects like MultiTool need to include things
+    # that are not specified in the project file, so their packages still require
+    # a .nuspec file.
+    foreach ($project in $Projects.NewApplication) {
+        New-NuGetPackageFromNuSpecFile $configuration $project $version
+    }
+}
+
+function Get-NuGetApiKey {
```

Comment:
>function [](start = 0, length = 8)

New code starts here. Note the hacky implementation, and see the comment. I don't think it's worth improving this until we put the key into the key vault.